include/pybind11/cast.h

```diff
@@ -434,11 +438,19 @@ struct string_caster {
     cast(const StringType &src, return_value_policy /* policy */, handle /* parent */) {
         const char *buffer = reinterpret_cast<const char *>(src.data());
         auto nbytes = ssize_t(src.size() * sizeof(CharT));
-        handle s = decode_utfN(buffer, nbytes);
-        if (!s) {
-            throw error_already_set();
+        if (std::is_same<CharT, uint8_t>::value
+#if defined(PYBIND11_CPP17)
+            || std::is_same<CharT, std::byte>::value
+#endif
```

Comment:
I'd prefer if the logic of what is a binary type was defined in one place, perhaps via splitting up the `is_std_char_type` into this:
```C++
template <typename CharT>
using is_std_binary_char_type = any_of<
#if defined(PYBIND11_CPP17)
                                       std::is_same<CharT, std::byte>,
#endif
                                       std::is_same<CharT, uint8_t>
                                       >;

template <typename CharT>
using is_std_char_type = any_of<std::is_same<CharT, char>, /* std::string */
#if defined(PYBIND11_HAS_U8STRING)
                                std::is_same<CharT, char8_t>, /* std::u8string */
#endif
                                std::is_same<CharT, char16_t>, /* std::u16string */
                                std::is_same<CharT, char32_t>, /* std::u32string */
                                std::is_same<CharT, wchar_t>,  /* std::wstring */
                                is_std_binary_char_type<CharT>
                                >;
```
and then here you can simplify to just:
```C++
        if (is_std_binary_char_type<CharT>::value)
            return PyBytes_FromStringAndSize(buffer, nbytes);
```

Reply:
Another random suggestion (entirely optional, some people might hate it, it just occurred to me as something that might clean up a bit):
```C++
// in detail/common.h:
template <typename T, typename... Ts>
using is_any_same = any_of<std::is_same<T, Ts>...>;
```
and then the `any_of<std::is_same<CharT, A>, std::is_same<CharT, B>, ...>` can become `is_any_same<CharT, A, B, ...>`.

I'd be surprised if there aren't a few other places in the code base that could be simplified using it.